Update continental extension cookbook

naliboff opened the pull request:
This PR updates to the continental extension cookbook to take advantage of new features in ASPECT and approaches published in recent papers. Additional updates will be required after #4204 and #4183 are merged, but I think this is a good start for now. FYI, an additional cookbook focusing on the rift to inversion will be forthcoming that uses elasticity, the EBA approximation, and other new features (student is currently working on these models).

* [X] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).
* [X] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

bangerth commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,144 @@
+set Nonlinear solver scheme                = single Advection, iterated defect correction Stokes
 set Nonlinear solver tolerance             = 1e-4
-set Max nonlinear iterations               = 10
+set Max nonlinear iterations               = 100
```
Do you actually need this?

tjhei replied:
I prefer this over silently continuing after 10 iterations...

naliboff replied:
Yes, on the first time step it requires quite a few nonlinear iterations (<5 after that).

bangerth commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,144 @@
-  set Initial adaptive refinement        = 0
-  set Initial global refinement          = 0
+  set Initial adaptive refinement        = 1
+  set Initial global refinement          = 2
```
Yes, it's nicer to create the mesh via refinement than via the initial X/Y repetitions.

bangerth commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,144 @@
+    set Number of cheap Stokes solver steps = 0
+  end
+  subsection Newton solver parameters
+    set Max pre-Newton nonlinear iterations      = 100
```
That's a lot of pre-Newton steps. In essence, you're saying that Newton doesn't work for this example. Is this the case?

naliboff replied:
Ah, thanks for catching that @bangerth! I previously used this parameter and `Nonlinear Newton solver switch tolerance` when  `iterated defect correction Stokes` was not available as a solver option and I wanted to make sure Newton iterations did not start. Removing both of these parameters.

bangerth commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,144 @@
+  set Run postprocessors on initial refinement = false
+end
+
+# Use the Eisenstat Walker method to automatically determine the
```
That's going to make @MFraters happy :-)

naliboff replied:
It has worked extremely well so far! I used it extensively in a series of recent 3D models.

bangerth commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,144 @@
-# Advecting the free surface vertically rather than
-# in the surface normal direction can result in a
-# more stable mesh when the deformation is large
+# Advecting the free surface using a normal, rather than vertical,
```
```suggestion
# Advect the free surface using a normal, rather than vertical,
```

bangerth commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -166,7 +214,7 @@ subsection Material model
 
     # Reference temperature and viscosity
-    set Reference temperature = 293
+    set Reference temperature = 273
```
That's pretty cold :-)

naliboff replied:
I actually don't recall why I made that change. Reasonable for the West Antartica Rift on consistent basis, but only 1 or 2 months a year in most rifts 😄. I'm actually wondering now how much that shift could affect dynamics long-term (e.g., changing the depth of the brittle-ductile transition)

bangerth replied:
Ha, there's a research question for you: How does surface temperature affect rift behavior!

bangerth commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -9,16 +9,16 @@ \subsubsection{Continental extension}
 
 \paragraph{Continental Extension}
-The 2D Cartesian model spans 400 (x) by 100 (y) km and has a finite element grid with uniform 2 km spacing. Unlike the crustal deformation cookbook (see Section~\ref{sec:cookbooks-crustal-deformation}, the mesh is not refined with time.
+The 2D Cartesian model spans 200 (x) by 100 (y) km and has a finite element grid with 1.25 and 2.5 km grid spacing, respectively,above and  below 50 km depth. This variation in grid spacing is achieved with a single initial adaptive refinement step using the minimum refinement function strategy. Unlike the crustal deformation cookbook (see Section~\ref{sec:cookbooks-crustal-deformation}, the mesh ...
```
```suggestion
The 2D Cartesian model spans 200 ($x$) by 100 ($y$) km and has a finite element grid with 1.25 and 2.5 km grid spacing, respectively, above and  below 50 km depth. This variation in grid spacing is achieved with a single initial adaptive refinement step using the minimum refinement function strategy. Unlike the crustal deformation cookbook (see Section~\ref{sec:cookbooks-crustal-deformation}, the mesh is not refined with time.
```

bangerth commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -38,29 +38,22 @@ \subsubsection{Continental extension}
 \end{align}
 
-In this example, specifying the top (\SI{273}{K}) and bottom temperature (\SI{1573}{K}), thermal conductivity of each layer and radiogenic heat production in each layer provides enough constraints to successively solve for the temperature and heat flux at the top of the lower crust and mantle.
+In this example, specifying the top (\SI{273}{K}) temperature, surface heat flow (55 $mW/m^2$), and thermal conductivity and radiogenic heat production of each layer provides enough constraints to successively solve for the temperature and heat flux at the top of the lower crust and mantle.
```
I think this should work, but please check:
```suggestion
In this example, specifying the top (\SI{273}{K}) temperature, surface heat flow (\SI{55}{\milli\Watt\per\meter^2}), and thermal conductivity and radiogenic heat production of each layer provides enough constraints to successively solve for the temperature and heat flux at the top of the lower crust and mantle.
```

bangerth commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -38,29 +38,22 @@ \subsubsection{Continental extension}
+In this example, specifying the top (\SI{273}{K}) temperature, surface heat flow (55 $mW/m^2$), and thermal conductivity and radiogenic heat production of each layer provides enough constraints to successively solve for the temperature and heat flux at the top of the lower crust and mantle.
 
-As noted above, the mechanically weak seed placed within the mantle localizes the majority of deformation onto two conjugate shear bands that propagate from the surface of the seed to the free surface. After 5 million years of extension background `stretching' is clearly visible in the strain-rate field, but deformation is still largely focused within the set of conjugate shear bands originating  ...
+As noted above, the initial zone of randomized plastic strain within the model center and strain softening of the friction and cohesion produces an initial pattern of distributed and unlocalized deformation across the zone of initial plastic strain (Fig.~\ref{fig:continental_extension_cookbook_0myr}). After 5 million years of extension distributed faulting is clearly evident in both the active an ...
```
```suggestion
As noted above, the initial zone of randomized plastic strain within the model center and strain softening of the friction and cohesion produces an initial pattern of distributed and unlocalized deformation across the zone of initial plastic strain (Fig.~\ref{fig:continental_extension_cookbook_0myr}). After 5 million years of extension, distributed faulting is clearly evident in both the active and finite deformation fields and surface topography over an approximately 100 km wide region (Fig.~\ref{fig:continental_extension_cookbook_5myr}). While deformation is distributed within this region, the fault system is clearly asymetric as the rate of deformation and accumulated brittle strain varies between faults. Indeed, localization onto the two conjugate rift-bounding border faults is evident from the active deformation field. Notably, deformation of the free surface near the fixed left and right walls is evident at 5 Myr. Continued distortion of the surface mesh near the lateral boundaries may lead to solver issues, which can be overcome by either widening the model or allowing the mesh to deform along these boundaries.  
```

bangerth commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -38,29 +38,22 @@ \subsubsection{Continental extension}
 
-With further extension for millions of years, significant crustal thinning and surface topography development should occur in response to displacement along the conjugate shear bands. However, given that the model only extends to 100 km depth, the simulation will not produce a realistic representation of continental breakup due to the lack of an upwelling asthenosphere layer. Indeed, numerical st ...
-as temperature variations associated with lithospheric thinning exert a first-order influence on the deformation patterns. As noted below, numerous additional parameters may also affect the temporal evolution of deformation patterns.
+With further extension for millions of years, significant crustal thinning and surface topography development should occur in response to displacement along the rift-bounding faults. However, given that the model only extends to 100 km depth, the simulation will not produce a realistic representation of continental breakup due to the lack of an upwelling asthenosphere layer. Indeed, numerical stu ...
```
Does `\cite[for example]{Bru14}` yield what you want? I think that puts the "for example" into the square brackets.

naliboff replied:
Yep, that is what I had intended.

bangerth replied:
That's an uncommon style, at least in my community. You would more often find it done like
```
  (e.g., Winkler boundary conditions, see for example [42])
```

bangerth commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -38,29 +38,22 @@ \subsubsection{Continental extension}
-\caption{\it Compositional field after \num{5e6} years of extension. The black line marks the 550 $\degree C$ isotherm.}
-\label{fig:continental_extension_cookbook_composition_5e6yr}
+\includegraphics[width=\textwidth]{cookbooks/continental_extension/doc/continental_extension_cookbook_5myr.png}
+\caption{\it Active (strain rate second invariant in units of $s^{-1}$) and finite (plastic) deformation after \num{5e6} years of extension. The white line marks the (\SI{893}{K}) isotherm (initial Moho temperature).}
```
```suggestion
\caption{\it Active (strain rate second invariant in units of $\si{s}^{-1}$) and finite (plastic) deformation after \num{5e6} years of extension. The white line marks the (\SI{893}{K}) isotherm (initial Moho temperature).}
```

bangerth commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -38,29 +38,22 @@ \subsubsection{Continental extension}
-\caption{\it Strain rate (in units of $s^{-1}$) after \num{5e6} years of extension. The black line marks the 550 $\degree C$ isotherm.}
-\label{fig:continental_extension_cookbook_strainrate_5e6yr}
+\includegraphics[width=\textwidth]{cookbooks/continental_extension/doc/continental_extension_cookbook_0myr.png}
+\caption{\it Initial active deformation field (strain rate second invariant in units of $s^{-1}$) and distribution of plastic strain. The white line marks the (\SI{893}{K}) isotherm (initial Moho temperature).}
```
```suggestion
\caption{\it Initial active deformation field (strain rate second invariant in units of $\si{s}^{-1}$) and distribution of plastic strain. The white line marks the (\SI{893}{K}) isotherm (initial Moho temperature).}
```

bangerth reviewed (commented):
I haven't tried to understand the details of the many parameters, but overall this looks quite nicely!

naliboff commented:
@bangerth - Thanks for the review! I've incorporated all your suggestions, and also made a number of other updates (changes to cookbook, change log, added more authors).

@anne-glerum @valemagni - Since you are now authors (based on all our discussions 😄), would you like to do a review as well?

anne-glerum commented:
Sure, I'll have a look now!

anne-glerum commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -9,16 +9,16 @@ \subsubsection{Continental extension}
 \lstinputlisting[language=prmfile]{cookbooks/continental_extension/doc/continental_extension_boundary_conditions.prm.out}
 
-Sections of the lithosphere with distinct properties are represented by compositional fields for the upper crust (20 km thick), lower crust (10 km thick) and mantle lithosphere (70 km thick). A mechanically weak seed within the mantle lithosphere helps localize deformation. Material (viscous flow law parameters, cohesion, internal friction angle) and thermodynamic properties for each compositiona ...
+Sections of the lithosphere with distinct properties are represented by compositional fields for the upper crust (20 km thick), lower crust (20 km thick) and mantle lithosphere (60 km thick). Material (viscous flow law parameters, cohesion, internal friction angle) and thermodynamic properties for each compositional field are based largely on previous numerical studies. Dislocation creep viscous  ...
```
```suggestion
Sections of the lithosphere with distinct properties are represented by compositional fields for the upper crust (20 km thick), lower crust (20 km thick) and mantle lithosphere (60 km thick). Material (viscous flow law parameters, cohesion, internal friction angle) and thermodynamic properties for each compositional field are based largely on previous numerical studies. Dislocation creep viscous flow parameters are taken from published deformation experiments for wet quartzite \cite{RB04}, wet anorthite \cite{RGWD06} and dry olivine \cite{HK04}. Additional compositional fields are used to track plastic strain and the non-initial plastic strain, with the latter value tracking the same quantity as the plastic strain absent the initial plastic strain values. As discussed further on, the plastic strain is used to soften (e.g., reduce) the friction and cohesion through time based on user-specified bounds and magnitudes. The initial randomized values of plastic strain in the model center localize distributed deformation in this region.
```

anne-glerum commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -38,29 +38,22 @@ \subsubsection{Continental extension}
+In this example, specifying the top (\SI{273}{K}) temperature, surface heat flow (\SI{55}{\milli\watt\per\meter^2}), and thermal conductivity and radiogenic heat production of each layer provides enough constraints to successively solve for the temperature and heat flux at the top of the lower crust and mantle.
 
-As noted above, the mechanically weak seed placed within the mantle localizes the majority of deformation onto two conjugate shear bands that propagate from the surface of the seed to the free surface. After 5 million years of extension background `stretching' is clearly visible in the strain-rate field, but deformation is still largely focused within the set of conjugate shear bands originating  ...
+As noted above, the initial zone of randomized plastic strain within the model center and strain softening of the friction and cohesion produces an initial pattern of distributed and unlocalized deformation across the zone of initial plastic strain (Fig.~\ref{fig:continental_extension_cookbook_0myr}). After 5 million years of extension, distributed faulting is clearly evident in both the active a ...
```
```suggestion
As noted above, the initial zone of randomized plastic strain within the model center and strain softening of the friction and cohesion produces an initial pattern of distributed and unlocalized deformation across the zone of initial plastic strain (Fig.~\ref{fig:continental_extension_cookbook_0myr}). After 5 million years of extension, distributed faulting is clearly evident in both the active and finite deformation fields and surface topography over an approximately 100 km wide region (Fig.~\ref{fig:continental_extension_cookbook_5myr}). While deformation is distributed within this region, the fault system is clearly asymmetric as the rate of deformation and accumulated brittle strain varies between faults. Localization onto the two conjugate rift-bounding border faults is evident from the active deformation field. Notably, deformation of the free surface near the fixed left and right walls is evident at 5 Myr. Continued distortion of the surface mesh near the lateral boundaries may lead to solver issues, which can be overcome by either widening the model or allowing the mesh to deform along these boundaries.  
```

anne-glerum commented on cookbooks/continental_extension/doc/continental_extension_geometry_mesh.prm:
```diff
@@ -1,15 +1,22 @@
+  subsection Minimum refinement function
+    set Coordinate system   = cartesian
+    set Variable names      = x,y
+    set Function expression = if ( y>=50e3 && x>=40.e3 && x<=160.e3, 4, 3)
```
Same comments apply here

anne-glerum commented on tests/continental_extension.prm:
```diff
@@ -15,38 +15,16 @@ include $ASPECT_SOURCE_DIR/cookbooks/continental_extension/continental_extension
 
 set Timing output frequency                = 100
 set End time                               = 1e4
+set Max nonlinear iterations               = 1
```
Isn't it possible to remove the temperature boundary conditions as well? I think they're the same as in the cookbook?

naliboff replied:
Yep!

anne-glerum reviewed (changes requested):
Great update @naliboff ! I made some comments on the text, but there is one thing that should prob be changed. The mesh refinement does only 3 additional refinements, but you're expecting 4 from what the minimum refinement function is saying.

bangerth commented:
Ping?

bangerth commented:
@naliboff Any news?

bangerth commented:
@naliboff Would you mind taking a look at the suggested changes above? Maybe we can still get this merged before the release.

naliboff commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,142 @@
 set CFL number                             = 0.5
+set Maximum time step                      = 20e3
 set Output directory                       = output-continental_extension
 set Timing output frequency                = 1
```
I typically do output every time step for the purposes of later debugging. It ends up being such a small component of the total run time that I have not felt the need to decrease the frequency yet.

naliboff commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,142 @@
 
-# Mesh refinement specifications (no mesh refinement,
-# but the coarse mesh is already 200x50, see above)
+# Adaptively refine the mesh to 1.25 km spacing above y=50 km
```
Updated!

naliboff commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -7,92 +7,142 @@
+  subsection Minimum refinement function
+    set Coordinate system   = cartesian
+    set Variable names      = x,y
+    set Function expression = if ( y>=50e3 && x>=40.e3 && x<=160.e3, 4, 3)
```
The above line does actually produce a spatially variable resolution, and effectively bumps up the minimum resolution to 2.5 km. I've increased the number of global refinements to 3, such that the equation now only modifies the resolution with the specified "high-resolution" box. The documentation has been updated accordingly as well.

naliboff commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -9,16 +9,16 @@ \subsubsection{Continental extension}
 
 \paragraph{Continental Extension}
-The 2D Cartesian model spans 400 (x) by 100 (y) km and has a finite element grid with uniform 2 km spacing. Unlike the crustal deformation cookbook (see Section~\ref{sec:cookbooks-crustal-deformation}, the mesh is not refined with time.
+The 2D Cartesian model spans 200 ($x$) by 100 ($y$) km and has a finite element grid with 1.25 and 2.5 km grid spacing, respectively, above and  below 50 km depth. This variation in grid spacing is achieved with a single initial adaptive refinement step using the minimum refinement function strategy. Unlike the crustal deformation cookbook (see Section~\ref{sec:cookbooks-crustal-deformation}, the ...
```
The PRM file has been updated such that the above text is correct.

naliboff commented:
@bangerth  - Apologies for taking so long to get back to this. I've addressed all of the remaining comments (thanks @anne-glerum for the thorough review!) and hopefully the testers are happy.

naliboff commented:
/rebuild

bangerth commented:
This PR is failing the "continental_extension" test. Can you check what is going on there?

naliboff commented:
@bangerth - The test is fixed, please let me know if this need any additional revisions.

anne-glerum reviewed (changes requested):
Hi John, thank you for addressing the comments. It's almost good to go!

anne-glerum commented on cookbooks/continental_extension/doc/continental_extension.tex:
```diff
@@ -9,16 +9,16 @@ \subsubsection{Continental extension}
 \lstinputlisting[language=prmfile]{cookbooks/continental_extension/doc/continental_extension_boundary_conditions.prm.out}
 
-Sections of the lithosphere with distinct properties are represented by compositional fields for the upper crust (20 km thick), lower crust (10 km thick) and mantle lithosphere (70 km thick). A mechanically weak seed within the mantle lithosphere helps localize deformation. Material (viscous flow law parameters, cohesion, internal friction angle) and thermodynamic properties for each compositiona ...
+Sections of the lithosphere with distinct properties are represented by compositional fields for the upper crust (20 km thick), lower crust (20 km thick) and mantle lithosphere (60 km thick). Material (viscous flow law parameters, cohesion, internal friction angle) and thermodynamic properties for each compositional field are based largely on previous numerical studies. Dislocation creep viscous  ...
```
localizes -> localize

naliboff commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -1,98 +1,148 @@
+# Initial values of different compositional fields
+# The upper crust (20 km thick), lower crust (20 km thick)
+# and mantle (60 km thick) are continuous horizontal layers
+# of varying thickness. The non initial plastic strain is set
```
Yep, you are right, this was confusing

naliboff commented on cookbooks/continental_extension/continental_extension.prm:
```diff
@@ -1,98 +1,148 @@
-  set Initial adaptive refinement        = 0
-  set Initial global refinement          = 0
+  set Initial adaptive refinement        = 1
+  set Initial global refinement          = 2
```
Fixed, thanks for catching that! I had changed changed in the geometry prm snippet, but not here.

naliboff commented:
@anne-glerum - Thanks for the additional round of revisions! I've hopefully addressed all of the comments, but let me know if I missed anything or you find anything else that needs to be updated.

anne-glerum reviewed (approved):
Thanks for the last round of revisions! This is ready to go!

anne-glerum commented:
Just one question, why did you change the test to do only 1 nonlinear solve per timestep?

naliboff commented:
@anne-glerum - sorry, just saw your last question! I changed the max nonlinear solves to 1 to decrease the test run time. If you think it would be useful to have multiple nonlinear iterations, I’ll update that in a follow-up PR. Thanks again for huge reviews! Would you or @bangerth like to merge?

anne-glerum commented:
Since you do 2 timesteps, I think it's okay to do just 1 nonlinear iterations. If it was only timestep 0, the solution would be dependent on the uniform initial strain rate, which doesn't really reflect the nonlinearity of the system. 
Thanks for all the updates!
